### PR TITLE
Setting all the optimizers to have useLocking = True

### PR DESCRIPTION
**Pull request by Craigacp**
I'm running into determinism issues when training models using TF-Java. This is one area which could be causing it, as in TF 2 all the optimizers have `useLocking=true`. We don't currently set this in TF-Java, and I'm worried the code path is degrading (as it says the behaviour may be undefined but faster with `useLocking=false`).

This doesn't resolve my non-determinism issue completely, but it seems a little better.

I've added a test to GradientDescentTest which checks that the models produced are identical. This test fails randomly, for two reasons, both of which are confusing. First it seems like when we fetch the weights from a model, we get a pointer to the weights, not a copy of them. This means that when they are trained the "copies" I'm saving out as the initialized ones are being updated. Second, and this is the real issue, the gradient updates can be different on identical models for identical data, for no apparent reason. I think this is happening somewhere in the C API, as I can't see where it could be happening in our code and the models use *identical GraphDefs*.

### CODE REVIEW

**Comment by karllessard** [comment]
@rnett , looks like the quick build is failing after merging #306 , can you please take a look at it?

**Comment by karllessard** [comment]
@Craigacp can you please rebase once more so we can give the quick-build another try?

**Comment by Craigacp** [comment]
Done.

**Comment by Craigacp** [comment]
Looks like everything passed this time.

File: tensorflow-framework/src/main/java/org/tensorflow/framework/optimizers/AdaGrad.java [hunk 1]
```diff
@@ -42,6 +43,9 @@
   public static final float LEARNING_RATE_DEFAULT = 0.001f;
   public static final float INITIAL_ACCUMULATOR_DEFAULT = 0.01f;
 
+  private static final ApplyAdagrad.Options[] opts = new ApplyAdagrad.Options[]{
+          ApplyAdagrad.updateSlots(true),ApplyAdagrad.useLocking(true)};
```
Comment (karllessard) [inline tensorflow-framework/src/main/java/org/tensorflow/framework/optimizers/AdaGrad.java:47]: Super nit: any formatter will probably complain about the missing space after a comma.

File: tensorflow-framework/src/test/java/org/tensorflow/framework/optimizers/GradientDescentTest.java [hunk 2]
```diff
@@ -97,4 +112,158 @@ public void testBasic() {
...
+                .feed("output", targetTensor)
+                .fetch(lossName)
+                .run().get(0);
+        postTrainingLoss[i] = lossVal.getFloat();
+        lossVal.close();
+      }
+    }
+
+    for (int i = 1; i < numRuns; i++) {
+      assertEquals(initialLoss[0],initialLoss[i]);
```
Comment (karllessard) [inline tensorflow-framework/src/test/java/org/tensorflow/framework/optimizers/GradientDescentTest.java:240]: Super nit: spaces after commas.

File: tensorflow-framework/src/test/java/org/tensorflow/framework/optimizers/GradientDescentTest.java [hunk 3]
```diff
@@ -97,4 +112,158 @@ public void testBasic() {
...
+        s.run(initName);
+
+        initialized.add(
+            s.runner()
+                .fetch(fcWeightName)
+                .fetch(fcBiasName)
+                .fetch(outputWeightName)
+                .fetch(outputBiasName)
+                .run());
+        System.out.println("Initialized - " + ndArrToString((TFloat32)initialized.get(i).get(3)));
```
Comment (karllessard) [inline tensorflow-framework/src/test/java/org/tensorflow/framework/optimizers/GradientDescentTest.java:207]: Can we avoid the verbosity in the unit test, aren't the equality check enough to validate? I'm fine just commenting out these `println`

**Review by karllessard — APPROVED** [review]
Thanks @Craigacp , I've approved it but left a few minor comments here and there, if you want to take a look

**Comment by Craigacp** [comment]
I'll clean up the test on Monday, and split it into two. There should be one test for the outputs returning a reference rather than a copy of the weights (as this might allow you to mutate the weights directly, which seems bad), and that test which just checks the gradient behaviour.

It's all conflated in that single test as I spent hours trying to figure out what was going on, so it has got all the print statements and other stuff necessary for me to track it down.

I'll clean up the formatting issues at the same time.

**Review by karllessard — APPROVED** [review]
